pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,282 @@
+
+	n.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, _ interface{}) {
+			n.Queue.Add(n.host)
+		},
+	})
+
+	go n.nodeInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.nodeInformer.HasSynced, n.kubevirtInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+
+	if !n.kvmController.NodeHasDevice(kvmPath) {
+		n.logger.Errorf("Node-labeller cannot work without KVM device.")
+		return true
+	}
+
+	err := n.run()
+
+	if err != nil {
+		n.logger.Errorf("node-labeller sync error encountered: %v", err)
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
+
+	//parse all informations
+	cpuModels, cpuFeatures = n.getCPUInfo()
+
+	//load node
+	nodeObj, exists, err := n.nodeInformer.GetStore().GetByKey(n.host)
+	if err != nil || !exists {
+		n.logger.Errorf("node-labeller cannot get node %s", n.host)
+		return err
+	}
+	var (
+		ok           bool
+		originalNode *v1.Node
+	)
+	if originalNode, ok = nodeObj.(*v1.Node); !ok {
+		n.logger.Errorf("node-labeller cannot convert node " + n.host)
+		return fmt.Errorf("Could not convert node " + n.host)
+	}
+
+	node := originalNode.DeepCopy()
+	//prepare new labels
+	newLabels := n.prepareLabels(cpuModels, n.hypervFeatures.items, cpuFeatures)
+	//remove old labeller labels
+	n.removeLabellerLabels(node)
+	//add new labels
+	n.addLabellerLabels(node, newLabels)
+
+	//patch node only if there is change in labels
+	err = n.patchNode(originalNode, node)
+	return err
+}
+
+func (n *NodeLabeller) patchNode(originalNode, node *v1.Node) error {
+	originalLabelsBytes, err := json.Marshal(originalNode.Labels)
+	if err != nil {
+		return err
+	}
+
+	originalAnnotationsBytes, err := json.Marshal(originalNode.Annotations)
+	if err != nil {
+		return err
+	}
+
+	labelsBytes, err := json.Marshal(node.Labels)
+	if err != nil {
+		return err
+	}
+
+	annotationsBytes, err := json.Marshal(node.Annotations)
+	if err != nil {
+		return err
+	}
+
+	if !reflect.DeepEqual(labelsBytes, originalLabelsBytes) {
+		patchTestLabels := fmt.Sprintf(`{ "op": "test", "path": "/metadata/labels", "value": %s}`, string(originalLabelsBytes))
+		patchTestAnnotations := fmt.Sprintf(`{ "op": "test", "path": "/metadata/annotations", "value": %s}`, string(originalAnnotationsBytes))
+		patchLabels := fmt.Sprintf(`{ "op": "replace", "path": "/metadata/labels", "value": %s}`, string(labelsBytes))
+		patchAnnotations := fmt.Sprintf(`{ "op": "replace", "path": "/metadata/annotations", "value": %s}`, string(annotationsBytes))
+		data := []byte(fmt.Sprintf("[ %s, %s ]", patchTestLabels, patchLabels))
```

Comment:
Can you create a `struct` for the json patch and then use `json.Marshal()` to serialize it? It would be safer than using a direct string manipulation. You could use an inline `struct`.